sunnypilot/selfdrive/controls/lib/dec/dec.py

```diff
@@ -0,0 +1,394 @@
+    if len(self.data) == 0:
+      return None
+    weighted_sum: float = float(np.dot(self.data, self.weights[-len(self.data):]))
+    weight_total: float = float(np.sum(self.weights[-len(self.data):]))
+    return weighted_sum / weight_total
+
+  def reset_data(self) -> None:
+    self.data = []
+
+
+class DynamicExperimentalController:
+  def __init__(self, params=None):
+    self._params = params or Params()
+    self._enabled: bool = self._params.get_bool("DynamicExperimentalControl")
+    self._active: bool = False
+    self._mode: str = 'acc'
+    self._frame: int = 0
+
+    # Use weighted moving average for filtering leads
+    self._lead_gmac = WeightedMovingAverageCalculator(window_size=LEAD_WINDOW_SIZE)
+    self._has_lead_filtered = False
+    self._has_lead_filtered_prev = False
+
+    self._slow_down_gmac = WeightedMovingAverageCalculator(window_size=SLOW_DOWN_WINDOW_SIZE)
+    self._has_slow_down = False
+
+    self._has_blinkers = False
+
+    self._slowness_gmac = WeightedMovingAverageCalculator(window_size=SLOWNESS_WINDOW_SIZE)
+    self._has_slowness = False
+
+    self._has_nav_instruction = False
+
+    self._dangerous_ttc_gmac = WeightedMovingAverageCalculator(window_size=DANGEROUS_TTC_WINDOW_SIZE)
+    self._has_dangerous_ttc = False
+
+    self._v_ego_kph = 0.
+    self._v_cruise_kph = 0.
+
+    self._has_lead = False
+
+    self._has_standstill = False
+    self._has_standstill_prev = False
+
+    self._sng_transit_frame = 0
+    self._sng_state = SNG_State.off
+
+    self._mpc_fcw_gmac = WeightedMovingAverageCalculator(window_size=MPC_FCW_WINDOW_SIZE)
+    self._has_mpc_fcw = False
+    self._mpc_fcw_crash_cnt = 0
+
+    self._set_mode_timeout = 0
+
+  @staticmethod
+  def _anomaly_detection(recent_data: list[float], threshold: float = 2.0, context_check: bool = True) -> bool:
+    """
+    Basic anomaly detection using standard deviation.
+    """
+    if len(recent_data) < 5:
+      return False
+    mean: float = float(np.mean(recent_data))
+    std_dev: float = float(np.std(recent_data))
+    anomaly: bool = bool(recent_data[-1] > mean + threshold * std_dev)
+
+    # Context check to ensure repeated anomaly
+    if context_check:
+      return np.count_nonzero(np.array(recent_data) > mean + threshold * std_dev) > 1
+    return anomaly
+
+  def _adaptive_slowdown_threshold(self) -> float:
+    """
+    Adapts the slow-down threshold based on vehicle speed and recent behavior.
+    """
+    adaptive_threshold: float = float(
+      interp(self._v_ego_kph, SLOW_DOWN_BP, SLOW_DOWN_DIST) * (1.0 + 0.05 * np.log(1 + len(self._slow_down_gmac.data)))
+    )
+    return adaptive_threshold
+
+  def _smoothed_lead_detection(self, lead_prob: float, smoothing_factor: float = 0.2):
+    """
+    Smoothing the lead detection to avoid erratic behavior.
+    """
+    self._has_lead_filtered = (1 - smoothing_factor) * self._has_lead_filtered + smoothing_factor * lead_prob
+    return self._has_lead_filtered > LEAD_PROB
```

Comment:
@rav4kumar `self._has_lead_filtered` is defined as `bool` when initialized:
https://github.com/sunnypilot/sunnypilot/blob/1f39c4ccfb0305d35f1ae14f0023980866a03ef4/sunnypilot/selfdrive/controls/lib/dec/dec.py#L121-L122

but in L183 it should be receiving a `float` instead:
https://github.com/sunnypilot/sunnypilot/blob/1f39c4ccfb0305d35f1ae14f0023980866a03ef4/sunnypilot/selfdrive/controls/lib/dec/dec.py#L179-L184

 Is this intentional? This is the same in the latest `dev-c3`: https://github.com/sunnypilot/sunnypilot/blob/167ef31602ef62abf567664338ad8ad3e13a30f4/selfdrive/controls/lib/sunnypilot/dynamic_experimental_controller.py#L177-L178